### PR TITLE
Player or allied force mk2

### PR DESCRIPTION
**Pull request by NickAragua**
This PR implements functionality that lets a scenario template be defined to have forces that are *either* RNG-supplied or player-supplied. On the scenario side, you can swap in a player unit back and forth with units in the appropriate template, similar to legacy AtB's "riot control" scenario. On the StratCon side, when generating "involuntary" scenarios or when deploying transported forces, eligible units in the player's TO&E automatically get subbed in.

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 1]
```diff
@@ -182,6 +186,73 @@ public static void generateScenariosForTrack(Campaign campaign, AtBContract cont
...
+                for (UUID unitID : scenario.getBackingScenario().getBotUnitTemplates().keySet()) {
+                    if (scenario.getBackingScenario().getBotUnitTemplates().get(unitID).getForceName().equals(sft.getForceName())) {
+                        unitCount++;
+                    }
+                }
+                
+                // or the units embedded in bot forces
+                for (BotForce botForce : scenario.getBackingScenario().getBotForceTemplates().keySet()) {
+                    if (scenario.getBackingScenario().getBotForceTemplates().get(botForce).getForceName().equals(sft.getForceName())) {
+                        unitCount += botForce.getEntityList().size();
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:209]:
```suggestion
                for (var botForce : scenario.getBackingScenario().getBotForceTemplates().entrySet()) {
                    if (gotForce.GetValue().getForceName().equals(sft.getForceName())) {
                        unitCount += botForce.getKey().getEntityList().size();
```

File: MekHQ/data/scenariotemplates/Defend Grounded Dropship.xml [hunk 2]
```diff
@@ -15,6 +15,59 @@
...
+            </successEffects>
+            <failureEffects>
+                <failureEffect>
+                    <effectType>ScenarioDefeat</effectType>
+                    <effectScaling>Fixed</effectScaling>
+                    <howMuch>2</howMuch>
+                </failureEffect>
+            </failureEffects>
+            <additionalDetails>
+                <additionalDetail>If the dropship cannot be deployed, this objective may be ignored.</additionalDetail>
```
Comment (Windchild292) [inline MekHQ/data/scenariotemplates/Defend Grounded Dropship.xml:39]:
```suggestion
                <additionalDetail>If the DropShip cannot be deployed, this objective may be ignored.</additionalDetail>
```

File: MekHQ/data/scenariotemplates/Low Atmo - Escort.xml [hunk 3]
```diff
@@ -15,6 +15,69 @@
...
+            <description>Destroy or rout 50% of the following force(s) and unit(s):</description>
+            <destinationEdge>NONE</destinationEdge>
+            <objectiveCriterion>ForceWithdraw</objectiveCriterion>
+            <percentage>50</percentage>
+            <timeLimitAtMost>true</timeLimitAtMost>
+            <timeLimitType>None</timeLimitType>
+        </scenarioObjective>
+        <scenarioObjective>
+            <associatedForceNames>
+                <associatedForceName>Allied Dropship</associatedForceName>
```
Comment (Windchild292) [inline MekHQ/data/scenariotemplates/Low Atmo - Escort.xml:48]:
```suggestion
                <associatedForceName>Allied DropShip</associatedForceName>
```

File: MekHQ/data/scenariotemplates/Defend Grounded Dropship.xml [hunk 4]
```diff
@@ -15,6 +15,59 @@
         <useStandardAtBSizing>true</useStandardAtBSizing>
         <widthScalingIncrement>5</widthScalingIncrement>
     </mapParameters>
+    <scenarioObjectives>
+        <scenarioObjective>
+            <associatedForceNames>
+                <associatedForceName>Dropship</associatedForceName>
```
Comment (Windchild292) [inline MekHQ/data/scenariotemplates/Defend Grounded Dropship.xml:21]:
```suggestion
                <associatedForceName>DropShip</associatedForceName>
```

File: MekHQ/data/scenariotemplates/Space - Blockade Run Escort.xml [hunk 5]
```diff
@@ -15,6 +15,69 @@
         <useStandardAtBSizing>false</useStandardAtBSizing>
         <widthScalingIncrement>5</widthScalingIncrement>
     </mapParameters>
+    <scenarioObjectives>
+        <scenarioObjective>
+            <associatedForceNames>
+                <associatedForceName>Escorted Dropship</associatedForceName>
```
Comment (Windchild292) [inline MekHQ/data/scenariotemplates/Space - Blockade Run Escort.xml:21]:
```suggestion
                <associatedForceName>Escorted DropShip</associatedForceName>
```

File: MekHQ/data/scenariotemplates/Space - Blockade Run.xml [hunk 6]
```diff
@@ -13,6 +15,35 @@
         <useStandardAtBSizing>false</useStandardAtBSizing>
         <widthScalingIncrement>5</widthScalingIncrement>
     </mapParameters>
+    <scenarioObjectives>
+        <scenarioObjective>
+            <associatedForceNames>
+                <associatedForceName>Enemy Dropships</associatedForceName>
```
Comment (Windchild292) [inline MekHQ/data/scenariotemplates/Space - Blockade Run.xml:21]:
```suggestion
                <associatedForceName>Enemy DropShips</associatedForceName>
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 7]
```diff
@@ -240,6 +241,16 @@ public void run() {
                         entity.setForceString(force.getFullMMName());
                     }
                     entities.add(entity);
+                    
+                    // if we've swapped this entity in for a bot-controlled unit, copy the bot controlled unit's
+                    // deployment parameters to this entity.
+                    if (scenario instanceof AtBDynamicScenario &&
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java:247]: brackets

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 8]
```diff
@@ -240,6 +241,16 @@ public void run() {
...
+                            ((AtBDynamicScenario) scenario).getPlayerUnitSwaps().
+                                containsKey(UUID.fromString(entity.getExternalIdAsString()))) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java:249]: We should probably null check the external id, as it is nullable IIRC

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java [hunk 9]
```diff
@@ -90,7 +112,7 @@ public void addForces(int forceID) {
         // loop through all player-supplied forces in the template, if there is one
         // assign the newly-added force to the first template we find
         if(template != null) {
-            for(ScenarioForceTemplate forceTemplate : template.scenarioForces.values()) {
+            for(ScenarioForceTemplate forceTemplate : template.getAllScenarioForces()) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java:115]:
```suggestion
            for (ScenarioForceTemplate forceTemplate : template.getAllScenarioForces()) {
```

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java [hunk 10]
```diff
@@ -183,12 +207,15 @@ public void setMapSize() {
...
     public void addBotForce(BotForce botForce, ScenarioForceTemplate forceTemplate) {
+        botForce.setTemplateName(forceTemplate.getForceName());
         super.addBotForce(botForce);
         botForceTemplates.put(botForce, forceTemplate);
+        botTemplateUnits.putIfAbsent(forceTemplate.getForceName(), new ArrayList<>());
 
         // put all bot units into the external ID lookup.
         for (Entity entity : botForce.getEntityList()) {
             getExternalIDLookup().put(entity.getExternalIdAsString(), entity);
+            botTemplateUnits.get(forceTemplate.getForceName()).add(UUID.fromString(entity.getExternalIdAsString()));
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java:218]: Null check for external id

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java [hunk 11]
```diff
@@ -431,10 +495,26 @@ protected void loadFieldsFromXmlNode(final Node wn, final Version version) throw
                 template = ScenarioTemplate.Deserialize(wn2);
             } else if (wn2.getNodeName().equalsIgnoreCase("finalized")) {
                 setFinalized(Boolean.parseBoolean(wn2.getTextContent().trim()));
+            } else if (wn2.getNodeName().equalsIgnoreCase(PLAYER_UNIT_SWAPS_ELEMENT)) {
+                //
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java:499]: Commented out nothing?

**Review by Windchild292 — COMMENTED** [review]
Few comments to add to the previous review.

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 12]
```diff
@@ -169,7 +170,7 @@ public void run() {
 
                 //minefields
                 client.getLocalPlayer().setNbrMFActive(scenario.getNumPlayerMinefields(Minefield.TYPE_ACTIVE));
-                client.getLocalPlayer().setNbrMFCommand(scenario.getNumPlayerMinefields(Minefield.TYPE_COMMAND_DETONATED));
+                //client.getLocalPlayer().setNbrMFCommand(scenario.getNumPlayerMinefields(Minefield.TYPE_COMMAND_DETONATED));
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/AtBGameThread.java:173]: No, we just don't have command detonated minefields implemented.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java [hunk 13]
```diff
@@ -458,4 +538,25 @@ public void clearAllForcesAndPersonnel(Campaign campaign) {
...
+    
+    /**
+     * Restores transient data that can be regenerated from already stored data
+     */
+    private void restoreTransientData() {
+        for (UUID unitID : botUnitTemplates.keySet()) {
+            String forceTemplateName = botUnitTemplates.get(unitID).getForceName();
+            botTemplateUnits.putIfAbsent(forceTemplateName, new ArrayList<>());
+            botTemplateUnits.get(forceTemplateName).add(unitID);
+        }
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenario.java:550]: Turns out we don't save these fields anyway, they're only used during scenario generation and then disappear.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 14]
```diff
@@ -2463,4 +2465,128 @@ private static boolean isPlanetOwner(AtBContract contract, LocalDate currentDate
...
+    
+    /**
+     * Given a player unit ID and a template name, if the player unit type matches  
+     * the template's unit type and the template generation method is PlayerOrAllied,
+     * take the first unit that we find in the given scenario that's a part of that
+     * template and "put it away".
+     */
+    public static void benchAllyUnit(UUID playerUnitID, String templateName, AtBDynamicScenario scenario) {
+        ScenarioForceTemplate destinationTemplate = null;
+        if (scenario.getTemplate().getScenarioForces().containsKey(templateName)) {
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java:2477]: Yes, they're all generated from templates.

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 8 above)
Comment (NickAragua) [reply MekHQ/src/mekhq/AtBGameThread.java:249]: I'm pretty sure every player and bot unit has an external ID. The bot units get one upon generation, while the player units always set the corresponding entity's external ID to the unit's ID.

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**